File: samples/python/s019_experimental_multires.py
#-----------------------------------------------------------------------
#Copyright 2013 Centrum Wiskunde & Informatica, Amsterdam
#
#Author: Daniel M. Pelt
#Contact: D.M.Pelt@cwi.nl
#Website: http://dmpelt.github.io/pyastratoolbox/
#
#
#This file is part of the Python interface to the
#All Scale Tomographic Reconstruction Antwerp Toolbox ("ASTRA Toolbox").
#
#The Python interface to the ASTRA Toolbox is free software: you can redistribute it and/or modify
#it under the terms of the GNU General Public License as published by
#the Free Software Foundation, either version 3 of the License, or
#(at your option) any later version.
#
#The Python interface to the ASTRA Toolbox is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#GNU General Public License for more details.
#
#You should have received a copy of the GNU General Public License
#along with the Python interface to the ASTRA Toolbox. If not, see <http://www.gnu.org/licenses/>.
#
#-----------------------------------------------------------------------

import astra
import numpy as np
from astra.experimental import do_composite_FP

astra.log.setOutputScreen(astra.log.STDERR, astra.log.DEBUG)

# low res part (voxels of 4x4x4)
vol_geom1 = astra.create_vol_geom(32, 16, 32, -64, 0, -64, 64, -64, 64)

# high res part (voxels of 1x1x1)
vol_geom2 = astra.create_vol_geom(128, 64, 128, 0, 64, -64, 64, -64, 64)


# Split the output in two parts as well, for demonstration purposes
angles1 = np.linspace(0, np.pi/2, 90, False)
angles2 = np.linspace(np.pi/2, np.pi, 90, False)
proj_geom1 = astra.create_proj_geom('parallel3d', 1.0, 1.0, 128, 192, angles1)
proj_geom2 = astra.create_proj_geom('parallel3d', 1.0, 1.0, 128, 192, angles2)

# Create a simple hollow cube phantom
cube1 = np.zeros((32,32,16))
cube1[4:28,4:28,4:16] = 1

cube2 = np.zeros((128,128,64))
cube2[16:112,16:112,0:112] = 1
cube2[33:97,33:97,4:28] = 0

vol1 = astra.data3d.create('-vol', vol_geom1, cube1)
vol2 = astra.data3d.create('-vol', vol_geom2, cube2)

proj1 = astra.data3d.create('-proj3d', proj_geom1, 0)
proj2 = astra.data3d.create('-proj3d', proj_geom2, 0)

# The actual geometries don't matter for this composite FP/BP case
projector = astra.create_projector('cuda3d', proj_geom1, vol_geom1)

do_composite_FP(projector, [vol1, vol2], [proj1, proj2])

proj_data1 = astra.data3d.get(proj1)
proj_data2 = astra.data3d.get(proj2)

# Display a single projection image
import pylab
pylab.gray()
pylab.figure(1)
pylab.imshow(proj_data1[:,0,:])
pylab.figure(2)
pylab.imshow(proj_data2[:,0,:])
pylab.show()


# Clean up. Note that GPU memory is tied up in the algorithm object,
# and main RAM in the data objects.
astra.data3d.delete(vol1)
astra.data3d.delete(vol2)
astra.data3d.delete(proj1)
astra.data3d.delete(proj2)
astra.projector3d.delete(projector)

File: samples/cpp/s019_experimental_multires.cpp
#include <cpp/creators.hpp>
#include <astra/CompositeGeometryManager.h>

using namespace std;
using namespace astra;

int main (int argc, char *argv[])
{
    int j,i,k;
    float32 *angles1, *angles2;
    float32 *cube1, *cube2;
    CVolumeGeometry3D *vol_geom1, *vol_geom2;
    CParallelProjectionGeometry3D *proj_geom1, *proj_geom2;
    CFloat32ProjectionData3DMemory *proj1, *proj2;
    CFloat32VolumeData3DMemory *vol1, *vol2;
    CCompositeGeometryManager *cgeomgr;
    CCudaProjector3D *proj;
    vector<CFloat32VolumeData3DMemory*> vols;
    vector<CFloat32ProjectionData3DMemory*> projs;

    // create phantom data (hollow cube)
    cube1 = new float32[16*32*32]();
    for(j=4;j<16;j++)
        for(i=4;i<28;i++)
            for(k=4;k<28;k++)
                cube1[j + i*16 + k*16*32] = 1.0;

    cube2 = new float32[64*128*128]();
    for(j=0;j<64;j++)
        for(i=16;i<112;i++)
            for(k=16;k<112;k++)
                cube2[j + i*64 + k*64*128] = 1.0;
    for(j=4;j<28;j++)
        for(i=33;i<97;i++)
            for(k=33;k<97;k++)
                cube2[j + i*64 + k*64*128] = 0.0;

    // volume geometry
    vol_geom1 = create_vol_geom_3d(32,16,32,-64,0,-64,64,-64,64);
    vol_geom2 = create_vol_geom_3d(128,64,128,0,64,-64,64,-64,64);

    // projection geometry
    angles1 = create_angles(0,M_PI_2,90);
    angles2 = create_angles(M_PI_2,M_PI,90);
    proj_geom1 = create_proj_geom_3d_parallel(1.0,1.0,128,192,90,angles1);
    proj_geom2 = create_proj_geom_3d_parallel(1.0,1.0,128,192,90,angles2);

    // data
    vol1 = create_data_3d_vol(vol_geom1,cube1);
    vol2 = create_data_3d_vol(vol_geom2,cube2);

    proj1 = create_data_3d_sino(proj_geom1,0.0);
    proj2 = create_data_3d_sino(proj_geom2,0.0);

    // store in vector
    vols.push_back(vol1);
    vols.push_back(vol2);

    projs.push_back(proj1);
    projs.push_back(proj2);

    // projector
    proj = create_projector_3d_cuda(proj_geom1,vol_geom1);

    // composite manager
    cgeomgr = new CCompositeGeometryManager();
    cgeomgr->doFP(proj,vols,projs);
    delete cgeomgr;

    // write projection data
    save_data_3d("sinogram_comp_1.dat",proj1);
    save_data_3d("sinogram_comp_2.dat",proj2);

    delete[] cube1;
    delete[] cube2;
    delete vol_geom1;
    delete vol_geom2;
    delete[] angles1;
    delete[] angles2;
    delete proj_geom1;
    delete proj_geom2;
    delete proj;
    delete vol1;
    delete vol2;
    delete proj1;
    delete proj2;
    
    return 0;
}
